### PR TITLE
boot: Add MCUBOOT_HW_KEY support for image encryption

### PR DESCRIPTION
**Pull request by DineshDK03**
Currently encryption supports only private key embed in mcuboot itself. To support MCUBOOT_HW_KEY for image encryption boot_retrieve_private_key() hook is added.

This hook helps retrieving private key from trusted sources like OTP, TPM.

### CODE REVIEW

File: boot/bootutil/src/encrypted.c
Comment (d3zd3z) [inline boot/bootutil/src/encrypted.c:519]: Make sure your editor is configured to insert spaces instead of tabs when working on mcuboot source. In this case, this should be 8 spaces rather than a tab.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
I like the idea of this change. I'd like to see if we can do this without introducing more ifdefs throughout the code. Perhaps we make the main code so that it always calls this function, and then, depending on the feature (ifdefs around whole functions are better than ones sprinkled through the code), we could just return a copy of the compiled in one. This could be conditionalized out if a given developer wishes to provide their own implementation.

File: boot/bootutil/src/encrypted.c
Comment (DineshDK03) [reply boot/bootutil/src/encrypted.c]: done

**Comment by DineshDK03** [comment]
@d3zd3z  please revisit.

**Comment by DineshDK03** [comment]
@d3zd3z Ping

File: boot/bootutil/src/encrypted.c
Comment (mingulov) [inline boot/bootutil/src/encrypted.c:441]: const struct means that this local `bootutil_enc_key` will remain {0} after a call to boot_retrieve_private_key

File: boot/cypress/MCUBootApp/keys.c
Comment (mingulov) [reply boot/cypress/MCUBootApp/keys.c:174]: Ok as I see - seems this pull request had been broken in the 2nd version, an original version had no such problem:
boot_retrieve_private_key(&bootutil_enc_key.key, &bootutil_enc_key.len);

Comment (DineshDK03) [reply boot/cypress/MCUBootApp/keys.c:174]: Thanks for the review and sorry, was on a long vacation, I will go through this latest version and update respectively

File: boot/bootutil/src/encrypted.c
Comment (mingulov) [inline boot/bootutil/src/encrypted.c:70]: I think it should be "struct bootutil_key"

Comment (DineshDK03) [reply boot/bootutil/src/encrypted.c]: sorry, yes it should be `struct bootutil_key bootutil_enc_key`. thanks for pointing it out. Since Zephyr doesn't support KW_ENCRYPT I haven't tested that part.

**Comment by jimmyzumthurm** [comment]
Any update on this ? We have similar solution currently in our software by patching MCUboot, but would highly appreciate if that would be officially available

**Comment by DineshDK03** [comment]
> Any update on this ? We have similar solution currently in our software by patching MCUboot, but would highly appreciate if that would be officially available

@jimmyzumthurm. I have already addressed the comments from @mingulov and @d3zd3z  and waiting for their acceptance to merge this PR.

**Comment by mingulov** [comment]
> > Any update on this ? We have similar solution currently in our software by patching MCUboot, but would highly appreciate if that would be officially available
> 
> @jimmyzumthurm. I have already addressed the comments from @mingulov and @d3zd3z and waiting for their acceptance to merge this PR.

I am not an official approver / maintainer for MCUboot, so unfortunately my acceptance would not help anyhow.

**Comment by DineshDK03** [comment]
@d3zd3z  @de-nordic @nordicjm please have a look.

**Comment by DineshDK03** [comment]
@d3zd3z ping

**Comment by nordicjm** [comment]
@sigvartmh

**Comment by DineshDK03** [comment]
@d3zd3z @sigvartmh @nordicjm @de-nordic please have a look, if all is good, kindly merge this PR.

File: boot/bootutil/src/encrypted.c
Comment (d3zd3z) [inline boot/bootutil/src/encrypted.c:70]: There should be a space after the comma.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
Other than the whitespace issue, this looks good now.

File: boot/bootutil/src/encrypted.c
Comment (DineshDK03) [reply boot/bootutil/src/encrypted.c]: fixed

**Comment by DineshDK03** [comment]
@d3zd3z Thanks for the review and fixed all the whitespace issues now.

**Review by d3zd3z — APPROVED** [review]

**Comment by DineshDK03** [comment]
@d3zd3z pushed changes to fix the CI. please approve to run the workflow jobs.

File: boot/mbed/app_enc_keys.c
Comment (d3zd3z) [inline boot/mbed/app_enc_keys.c:76]: This line, and the ones below, doesn't actually do anything. It it just copying the address of the key to a local variable. I suspect what you actually want is for the parameter to be `struct bootutil_key **private_key`.

**Review by d3zd3z — CHANGES_REQUESTED** [review]

File: boot/mbed/app_enc_keys.c
Comment (DineshDK03) [reply boot/mbed/app_enc_keys.c]: fixed it now. please trigger CI

Comment (DineshDK03) [reply boot/mbed/app_enc_keys.c]: @d3zd3z CI passed. I hope we can merge this now.

**Comment by DineshDK03** [comment]
@d3zd3z please have a look, all the comments addressed and waiting to be merged.

File: boot/bootutil/include/bootutil/enc_key.h
Comment (davidvincze) [inline boot/bootutil/include/bootutil/enc_key.h:58]: Would you mind renaming this function to `boot_enc_retrieve_priv_key` or similar?
It's for avoiding ambiguity - one might think that it's retrieving the private key used for image signing (there's already a function called `boot_retrieve_public_key_hash`).

File: boot/cypress/MCUBootApp/keys.c
Comment (davidvincze) [inline boot/cypress/MCUBootApp/keys.c:178]:
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

File: boot/mbed/app_enc_keys.c
Comment (davidvincze) [inline boot/mbed/app_enc_keys.c:80]:
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

File: boot/zephyr/keys.c
Comment (davidvincze) [inline boot/zephyr/keys.c:97]:
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

File: ci/mynewt_keys/enc_kw/src/keys.c
Comment (davidvincze) [inline ci/mynewt_keys/enc_kw/src/keys.c:39]:
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

File: ci/mynewt_keys/enc_rsa/src/keys.c
Comment (davidvincze) [inline ci/mynewt_keys/enc_rsa/src/keys.c:137]:
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

File: sim/mcuboot-sys/csupport/keys.c
Comment (davidvincze) [inline sim/mcuboot-sys/csupport/keys.c:339]:
```suggestion
#endif /* !MCUBOOT_HW_KEY && MCUBOOT_ENC_IMAGES */
```

**Comment by davidvincze** [comment]
@DineshDK03 please add a top commit that adds a release note snippet for this feature:
https://github.com/mcu-tools/mcuboot/blob/main/docs/SubmittingPatches.md#release-notes
Thank you!

File: sim/mcuboot-sys/csupport/keys.c
Comment (davidvincze) [inline sim/mcuboot-sys/csupport/keys.c:333]:
```suggestion
int boot_enc_retrieve_private_key(struct bootutil_key **private_key)
```

File: boot/bootutil/include/bootutil/enc_key.h
Comment (davidvincze) [reply boot/bootutil/include/bootutil/enc_key.h]: The renaming is still missing in one file. I've added a comment to it.

File: sim/mcuboot-sys/csupport/keys.c
Comment (DineshDK03) [reply sim/mcuboot-sys/csupport/keys.c]: missed it, updated now, thanks.

**Review by davidvincze — APPROVED** [review]

**Comment by davidvincze** [comment]
@d3zd3z any other comments on this PR or is it good to go?

**Pull request merged**
